[Doc] Update Serve architecture doc for 2.0 (#26861)
ray-project/ray: doc/source/serve/architecture.md

This section should help you:

- Get an overview of how each component in Serve works
- Understand the different types of actors that make up a Serve instance

> % Figure source: https://docs.google.com/drawings/d/1jSuBN5dkSj2s9-0eGzlU_ldsRa3TsswQUZM-cMQ29a0/edit?usp=sharing

Reviewer: figure need to be updated.

This section should help you:

- Get an overview of how each component in Serve works
- Understand the different types of actors that make up a Serve instance

% Figure source: https://docs.google.com/drawings/d/1jSuBN5dkSj2s9-0eGzlU_ldsRa3TsswQUZM-cMQ29a0/edit?usp=sharing

```{image} architecture-2.0.svg
:align: center
:width: 600px
```

## High-Level View

Serve runs on Ray and utilizes [Ray actors](actor-guide).

There are three kinds of actors that are created to make up a Serve instance:

- **Controller**: A global actor unique to each Serve instance that manages
  the control plane. The Controller is responsible for creating, updating, and
  destroying other actors. Serve API calls like creating or getting a deployment
  make remote calls to the Controller.
- **HTTP Proxy**: By default there is one HTTP proxy actor on the head node. This actor runs a [Uvicorn](https://www.uvicorn.org/) HTTP
  server that accepts incoming requests, forwards them to replicas, and
  responds once they are completed.  For scalability and high availability,

>   you can also run a proxy on each node in the cluster via the `location` field of `http_options`: {ref}`core-apis`

Reviewer: ```suggestion
  you can also run a proxy on each node in the cluster via the `location` field of [`http_options`](core-apis)
```

  the control plane. The Controller is responsible for creating, updating, and
  destroying other actors. Serve API calls like creating or getting a deployment
  make remote calls to the Controller.

> - HTTP Proxy: There is one HTTP proxy actor on the head node (the location and number of proxies is configurable via the `location` field of `http_options`: {ref}`core-apis`). This actor runs a [Uvicorn](https://www.uvicorn.org/) HTTP

Reviewer 2: Specify that _by default_ there's one on the head node. For scalability and HA, it can be run on every node.

- Worker Replicas: Actors that actually execute the code in response to a
  request. For example, they may contain an instantiation of an ML model. Each
  replica processes individual requests from the HTTP proxy (these may be batched
  by the replica using `@serve.batch`, see the [batching](serve-batching) docs).

## Lifetime of a Request

When an HTTP request is sent to the HTTP proxy, the following things happen:

1. The HTTP request is received and parsed.
2. The correct deployment associated with the HTTP URL path is looked up. The
  request is placed on a queue.
3. For each request in a deployment's queue, an available replica is looked up in round-robin fashion
  and the request is sent to it. If there are no available replicas (i.e. there
  are more than `max_concurrent_queries` requests outstanding at each replica), the request
  is left in the queue until a replica becomes available.

Each replica maintains a queue of requests and executes requests one at a time, possibly
using `asyncio` to process them concurrently. If the handler (the deployment function or the `__call__` method of the deployment class) is declared with `async def`, the replica will not wait for the

> handler to run.  Otherwise, the replica will block until the handler returns.

Reviewer: Not "the replica will not wait for the handler to run." but "the replica will run the asyncio methods concurrently with other invocations in the same event loop.

When making a request via [ServeHandle](serve-handle-explainer) instead of HTTP, the request is placed on a queue in the ServeHandle, and we skip to step 3 above.

(serve-ft-detail)=

## Fault tolerance

Application errors like exceptions in your model evaluation code are caught and
wrapped. A 500 status code will be returned with the traceback information. The
replica will be able to continue to handle requests.

Machine errors and faults will be handled by Ray Serve as follows:

- When replica actors fail, the Controller actor will replace them with new ones.
- When the HTTP proxy actor fails, the Controller actor will restart it.
- When the Controller actor fails, Ray will restart it.

> - When the Ray cluster fails, Ray Serve cannot recover.

Reviewer: Ray Serve replies on Kubernetes to recover.

Reviewer 2: please update that in kubernetes the operator will recover it.

also add a note that GCS HA can be used to avoid the cluster failure. ask @iycheng where we can link to relevant docs.

Author: Does the kubernetes note apply just to failed worker nodes, or to a failed cluster?  Also, this is referring specifically to Kuberay, right?

Reviewer 2: A failed cluster as well. And yes, specific to kuberay.

  the control plane. The Controller is responsible for creating, updating, and
  destroying other actors. Serve API calls like creating or getting a deployment
  make remote calls to the Controller.

> - HTTP Proxy: By default there is one HTTP proxy actor on the head node. This actor runs a [Uvicorn](https://www.uvicorn.org/) HTTP

Reviewer 3: Nit: Can we use bold font for the three types of actors (Controller, HTTP Proxy, and replicas), so they stand out from the definitions?

  server that accepts incoming requests, forwards them to replicas, and
  responds once they are completed.  For scalability and high availability,
  you can also run a proxy on each node in the cluster via the `location` field of `http_options`: {ref}`core-apis`

> - Worker Replicas: Actors that actually execute the code in response to a

Reviewer 3: `Worker` is a somewhat overloaded term throughout Ray. Can we call these `Deployment Replicas` throughout the documentation (including the diagram) instead?

Alternatively, especially if this term is used in more than just the architecture page, we can make this change in a future PR.

Author: Good point, I've just changed it to call them replicas since that's what we call them elsewhere

(serve-ft-detail)=

## Fault tolerance

Application errors like exceptions in your model evaluation code are caught and
wrapped. A 500 status code will be returned with the traceback information. The
replica will be able to continue to handle requests.

Machine errors and faults will be handled by Ray Serve as follows:

- When replica actors fail, the Controller actor will replace them with new ones.
- When the HTTP proxy actor fails, the Controller actor will restart it.
- When the Controller actor fails, Ray will restart it.
- When using Kuberay ({ref}`kuberay-index`), the Kuberay Operator will recover crashed nodes.
- When the Ray cluster fails, Ray Serve cannot recover.

When a machine hosting any of the actors crashes, those actors will be automatically restarted on another
available machine. All data in the Controller (routing policies, deployment
configurations, etc) is checkpointed to the Ray Global Control Store (GCS) on the head node. Transient data in the
router and the replica (like network connections and internal request
queues) will be lost for this kind of failure.

(serve-autoscaling-architecture)=

## Ray Serve Autoscaling

> Ray Serve's autoscaling feature automatically increases or decreases a deployment's number of replicas based on its load.

Reviewer 3: Can we mention somewhere in this section that the autoscaler runs in the Serve controller?

Author: Good point, added, and I'll add this in the diagram as well

![pic](https://raw.githubusercontent.com/ray-project/images/master/docs/serve/autoscaling.svg)

- Each ServeHandle and each replica periodically pushes its metrics to the autoscaler.

> - For each deployment, the autoscaler uses ServeHandle queue metrics and replicas queries metrics to make a decision whether to scale up or down the number of replicas.

Reviewer 3: ```suggestion
- For each deployment, the autoscaler periodically checks ServeHandle queues and in-flight queries on replicas to decide whether or not to scale the number of replicas.
```

> - Each ServeHandle continues to poll the updated group of replicas from the controller. Upon discovery of the new replicas, it will send any buffered or new queries to the replica until `max_concurrent_queries` is reached.

Reviewer 3: ```suggestion
- Each ServeHandle continuously polls the controller to check for new deployment replicas. Whenever new replicas are discovered, it will send any buffered or new queries to the replica until `max_concurrent_queries` is reached.
```
This sounds like `ServeHandles` briefly stop doing round-robin routing and instead opt to fill up new replicas first before distributing the workload. Is that true?

Author: I think that's only true if max_concurrent_queries has been reached on all the existing replicas, which is likely the case if we're autoscaling up.  That's basically a side effect of round robin with the constraint of max_concurrent_queries. 
 I'll update the doc to clarify

:::{note}

> When the controller dies, the client will still be able to send queries, but autoscaling will be paused. When the controller recovers, the autoscaling will resume, but all previous metrics collected will be lost.

Reviewer 3: Could we use a different word than client here? It's not totally clear on a first glance whether we mean the end user, the user managing the Serve deployments, the Serve controller client, or ServeHandles in general here.

Author: Good point, edited to "requests can still be sent via HTTP and ServeHandles"

:::

## Ray Serve API Server

Ray Serve provides a [CLI](serve-cli) for managing your Ray Serve instance, as well as a [REST API](serve-rest-api).

> The Ray Serve API server that responds to these REST requests is a module that is automatically loaded when Ray starts, and runs on the head node.

Reviewer 3: ```suggestion
Each node in your Ray cluster provides a Serve REST API server that can connect to Serve and respond to Serve REST requests
```

The Ray Serve REST API server isn't in the dashboard anymore. As part of the HA effort, the server got moved to the dashboard agent. One copy of the agent runs on each node in the Ray cluster. The agent is initialized lazily, so when a node receives a REST request for the first time, it initializes the agent.

Author: Ah good catch, I'll update the diagram too.

Serve utilizes Ray’s [shared memory object store](plasma-store) and in process memory
store. Small request objects are directly sent between actors via network

> call. Larger request objects (100KiB+) are written to the object store and the replica can read them via zero-copy read (TODO: doesn't this only apply to numpy arrays?).

Author: Doesn't zero-copy read only apply to numpy arrays?

Author: @simon-mo do you know how this works?  I thought only numpy arrays had zero-copy read, not arbitrary Ray objects.
